sound/soc/sof/ipc4-topology.c

```diff
@@ -1339,16 +1358,29 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 		copier_data = &ipc4_copier->data;
 		available_fmt = &ipc4_copier->available_fmt;
 		if (dir == SNDRV_PCM_STREAM_CAPTURE) {
+			unsigned int fmt_cfg;
+
 			format_list_to_search = available_fmt->output_pin_fmts;
 			format_list_count = available_fmt->num_output_formats;
 
 			/*
-			 * modify the input params for the dai copier as it only supports
-			 * 32-bit always
+			 * Sample format for capture pipeline is
+			 * defined here and it should be selected
+			 * based on what is configured in topology.
+			 * Here we pick the format from
+			 * output_pin_fmts[0]. This is always defined,
+			 * and we don't have the required context to
+			 * prioritize in case out_audio_fmt[1..N]
+			 * would define a different sample type.
 			 */
 			fmt = hw_param_mask(pipeline_params, SNDRV_PCM_HW_PARAM_FORMAT);
 			snd_mask_none(fmt);
-			snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+			fmt_cfg = available_fmt->output_pin_fmts[0].audio_fmt.fmt_cfg;
```

Comment:
@kv2019i this change to use the output format in topology is only because we only have 32-bit support in the topology today right? In theory we should just be able to use the fe_params for capture and pipeline_params for playback right?

Reply:
@ranj063 We don't want the pipeline to be 16bit just because FE is 16bit. This should be configured in topology (e.g. if you want higher precision processing in the processing pipeline). I.e. topology author can choose where precision is reduced in the pipeline.

Reply:
@kv2019i if the pipeline requires a different format for processing, the copier will do the conversion no? Of course,.this will avoid the conversion.

Reply:
@ranj063 Right but we are configuring the first copier here and as many components can support multiple formats, if we set S16 here, the pipeline will then be S16. There are multiple possible ways to configure the pipeline in these cases, but I think both a) configure always to S32 (current model), or b) locking to FE, are too simple policies.